eventstream.go

```diff
@@ -6,215 +6,236 @@ import (
+		for ref, items := range e.aggregateTypes {
+			e.aggregateTypes[ref] = clean(items)
 		}
 	}
-	s.subF = func() {
-		e.lock.Lock()
-		defer e.lock.Unlock()
+	e.lock.Lock()
+	defer e.lock.Unlock()
 
-		for _, a := range aggregates {
-			name := reflect.TypeOf(a).Elem().Name()
-			root := a.Root()
-			ref := fmt.Sprintf("%s_%s", root.path(), name)
+	for _, a := range aggregates {
+		name := reflect.TypeOf(a).Elem().Name()
+		root := a.Root()
+		ref := fmt.Sprintf("%s_%s", root.path(), name)
 
-			// adds one more function to the aggregate
-			e.aggregateTypes[ref] = append(e.aggregateTypes[ref], &s)
-		}
+		// adds one more function to the aggregate
+		e.aggregateTypes[ref] = append(e.aggregateTypes[ref], &s)
 	}
 	return &s
 }
 
-// SubscriberSpecificEvent bind the f function to be called on specific events
-func (e *EventStream) SubscriberSpecificEvent(f func(e Event), events ...interface{}) *Subscription {
-	s := Subscription{
-		f: f,
+// Event subscribe on specific application defined events based on type referencing.
+func (e *EventStream) Event(f func(e Event), events ...interface{}) *subscription {
+	s := subscription{
+		eventF: f,
 	}
-	s.unsubF = func() {
+	s.close = func() {
 		e.lock.Lock()
 		defer e.lock.Unlock()
-
-		for _, event := range events {
-			t := reflect.TypeOf(event)
-			for i, sub := range e.specificEvents[t] {
-				if &s == sub {
-					e.specificEvents[t] = append(e.specificEvents[t][:i], e.specificEvents[t][i+1:]...)
-					break
-				}
-			}
+		s.eventF = nil
+		// clean all evenF functions that are nil
+		for ref, items := range e.specificEvents {
+			e.specificEvents[ref] = clean(items)
 		}
 	}
-	s.subF = func() {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	for _, event := range events {
+		ref := reflect.TypeOf(event)
+		// adds one more property to the event type
+		e.specificEvents[ref] = append(e.specificEvents[ref], &s)
+	}
+	return &s
+}
+
+// Name subscribe to aggregate name combined with event names. The Name subscriber makes it possible to subscribe to
+// events event if the aggregate and event types are within the current application context.
+func (e *EventStream) Name(f func(e Event), aggregate string, events ...string) *subscription {
+	s := subscription{
+		eventF: f,
+	}
+	s.close = func() {
 		e.lock.Lock()
 		defer e.lock.Unlock()
-
-		// subscribe to specified events
-		for _, event := range events {
-			t := reflect.TypeOf(event)
-			// adds one more property to the event type
-			e.specificEvents[t] = append(e.specificEvents[t], &s)
+		s.eventF = nil
+		// clean all evenF functions that are nil
+		for ref, items := range e.names {
+			e.names[ref] = clean(items)
 		}
 	}
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	for _, event := range events {
+		ref := aggregate + "_" + event
+		e.names[ref] = append(e.names[ref], &s)
+	}
 	return &s
 }
+
+// removes subscriptions with event function equal to nil
+func clean(items []*subscription) []*subscription {
```

Comment:
Doesn't this function mutate `items` inplace, thus its not necessary to return the slice?
It would be nice with a functional approach, like so:

```go
func clean(items []*subscription) []*subscription {
    subs := []*subscription{}
    for _, s := range items {
        if s.eventF != nil {
            is = append(subs, s)
        }
    }
    return subs
```

Reply:
I thing the slice is inserted by value and will not mutate.